pulsar-client-cpp/lib/PublisherStatsImpl.cc

```diff
@@ -0,0 +1,129 @@
+
+    std::string PublisherStatsImpl::latencyToString(const LatencyAccumulator& obj) {
+        boost::accumulators::detail::extractor_result<LatencyAccumulator, boost::accumulators::tag::extended_p_square>::type latencies = boost::accumulators::extended_p_square(obj);
+        std::stringstream os;
+        os <<   "Latencies [ 50pct: " << latencies[0] / 1e3 << "ms" <<
+                ", 90pct: " << latencies[1] / 1e3 << "ms" <<
+                ", 99pct: " << latencies[2] / 1e3 << "ms" <<
+                ", 99.9pct: " << latencies[3] / 1e3 << "ms" <<
+                "]";
+        return os.str();
+    }
+
+    double PublisherStatsImpl::differenceInMicros(timespec& start, timespec& end) {
+        return (end.tv_sec - start.tv_sec) * 1e6 + (end.tv_nsec - start.tv_nsec) / 1e3;
+    }
+
+    PublisherStatsImpl::PublisherStatsImpl(std::string producerStr, DeadlineTimerPtr timer, uint64_t statsIntervalInSeconds)
+    : numMsgsSent_(0),
+    numBytesSent_(0),
+    numAcksReceived_(0),
+    totalMsgsSent_(0),
+    totalBytesSent_(0),
+    totalAcksReceived_(0),
+    timer_(timer),
+    producerStr_(producerStr),
+    statsIntervalInSeconds_(statsIntervalInSeconds),
+    mutex_(),
+    latencyAccumulator_(boost::accumulators::tag::extended_p_square::probabilities = probs),
+    totalLatencyAccumulator_(boost::accumulators::tag::extended_p_square::probabilities = probs) {
+        timer_->expires_from_now(boost::posix_time::seconds(statsIntervalInSeconds_));
+        timer_->async_wait(
+                boost::bind(&pulsar::PublisherStatsImpl::flushAndReset, this,
+                        boost::asio::placeholders::error));
+    }
+
+    void PublisherStatsImpl::printStats() {
+        LOG_INFO(*this);
+    }
+
+    void PublisherStatsImpl::flushAndReset(const boost::system::error_code& ec) {
+        if (ec) {
+            LOG_DEBUG("Ignoring timer cancelled event, code[" << ec <<"]");
+            return;
+        }
+        {
+            Lock lock(mutex_);
+            printStats();
+            numMsgsSent_ = 0;
+            numBytesSent_ = 0;
+            numAcksReceived_ = 0;
+            numSendFailedMap_.clear();
+            latencyAccumulator_ = LatencyAccumulator(boost::accumulators::tag::extended_p_square::probabilities = probs);
+        }
+        timer_->expires_from_now(boost::posix_time::seconds(statsIntervalInSeconds_));
+        timer_->async_wait(
+                boost::bind(&pulsar::PublisherStatsImpl::flushAndReset, this,
+                        boost::asio::placeholders::error));
+    }
+
+    void PublisherStatsImpl::messageSent(const Message& msg) {
+        Lock lock(mutex_);
+        numMsgsSent_++;
+        totalMsgsSent_++;
+        numBytesSent_ += msg.getLength();
+        totalBytesSent_ += msg.getLength();
+    }
+
+    void PublisherStatsImpl::messageReceived(Result& res, timespec& publishTime) {
+        timespec currentTime;
+        clock_gettime(CLOCK_REALTIME, &currentTime);
```

Comment:
i didn't realize we were using boost's posix time library. given that, consider using ```microsec_clock::universal_time()``` instead of ```clock_gettime()```